File: apps/api/src/lib/socket.ts
import { Server as SocketIOServer, Socket } from 'socket.io';
import type { Server as HTTPServer } from 'http';
import jwt from 'jsonwebtoken';
import { env } from '../config/env.js';
import { logger } from './logger.js';

export interface SocketUserPayload {
  userId?: string;
  orgId: string;
  /** Present for internal users only; customer tokens never carry a role. */
  role?: string;
  /** Token audience: 'internal' | 'customer'. Legacy tokens without `typ` are inferred. */
  typ?: 'internal' | 'customer';
  /** Customer tokens are scoped to an explicit allow-list of quotation ids. */
  quotationIds?: string[];
}

let io: SocketIOServer | null = null;

const ALLOWED_SOCKET_ORIGINS = (env.SOCKET_CORS_ORIGINS ?? 'http://localhost,http://localhost:5173,http://localhost:5174')
  .split(',')
  .map((o) => o.trim())
  .filter(Boolean);

export function initSocketIO(httpServer: HTTPServer): SocketIOServer {
  io = new SocketIOServer(httpServer, {
    cors: {
      origin: ALLOWED_SOCKET_ORIGINS,
      methods: ['GET', 'POST'],
    },
    path: '/socket.io',
  });

  // JWT Authentication & audience verification
  io.use((socket: Socket, next: (err?: Error) => void) => {
    try {
      const token =
        socket.handshake.auth?.token ||
        (socket.handshake.headers.authorization?.startsWith('Bearer ')
          ? socket.handshake.headers.authorization.slice(7)
          : null);

      if (!token) {
        return next(new Error('Authentication token required'));
      }

      const decoded = jwt.verify(token, env.JWT_SECRET) as any;
      if (!decoded || !decoded.org_id) {
        return next(new Error('Invalid token or missing organization context'));
      }

      // Super Admin tokens are platform-scoped: they have no tenant context
      // and must not join any org room.
      if (decoded.typ === 'super_admin' || decoded.role === 'super_admin') {
        return next(new Error('Super Admin tokens cannot open realtime connections'));
      }

      const typ: 'internal' | 'customer' =
        decoded.typ === 'customer' ? 'customer' : decoded.role ? 'internal' : 'customer';

      if (typ === 'customer' && decoded.role) {
        return next(new Error('Malformed customer token: internal role claims are not permitted'));
      }

      socket.data.user = {
        userId: decoded.sub || decoded.userId,
        orgId: decoded.org_id,
        typ,
        ...(typ === 'internal' ? { role: decoded.role } : {}),
        ...(typ === 'customer' ? { quotationIds: Array.isArray(decoded.quotation_ids) ? decoded.quotation_ids : [] } : {}),
      } as SocketUserPayload;

      next();
    } catch (err: any) {
      logger.warn({ err: err.message }, 'Socket.IO authentication failed');
      next(new Error('Unauthorized socket connection'));
    }
  });

  io.on('connection', (socket: Socket) => {
    const user = socket.data.user as SocketUserPayload;

    if (user.typ === 'internal') {
      // Internal users join the full tenant room and may subscribe to any
      // quotation room within their org.
      const orgRoom = `org:${user.orgId}`;
      socket.join(orgRoom);
      logger.info({ socketId: socket.id, orgId: user.orgId, role: user.role }, 'Internal socket joined org room');

      socket.on('quote:join', (quotationId: string) => {
        if (typeof quotationId === 'string' && quotationId.trim()) {
          const quoteRoom = `org:${user.orgId}:quote:${quotationId}`;
          socket.join(quoteRoom);
          logger.debug({ socketId: socket.id, room: quoteRoom }, 'Socket joined quote room');
        }
      });

      socket.on('quote:leave', (quotationId: string) => {
        if (typeof quotationId === 'string' && quotationId.trim()) {
          const quoteRoom = `org:${user.orgId}:quote:${quotationId}`;
          socket.leave(quoteRoom);
          logger.debug({ socketId: socket.id, room: quoteRoom }, 'Socket left quote room');
        }
      });
    } else {
      // Customer tokens never receive the org-wide event stream. They are
      // pinned to the quote rooms listed in their JWT allow-list only.
      const allowed = new Set(user.quotationIds ?? []);
      for (const quotationId of allowed) {
        socket.join(`org:${user.orgId}:quote:${quotationId}`);
      }
      logger.info(
        { socketId: socket.id, orgId: user.orgId, scopedQuotes: allowed.size },
        'Customer socket joined scoped quote rooms only'
      );

      socket.on('quote:join', (quotationId: string) => {
        if (typeof quotationId === 'string' && allowed.has(quotationId)) {
          socket.join(`org:${user.orgId}:quote:${quotationId}`);
        } else {
          logger.warn({ socketId: socket.id, quotationId }, 'Customer attempted to join unauthorized quote room');
        }
      });

      socket.on('quote:leave', (quotationId: string) => {
        if (typeof quotationId === 'string' && allowed.has(quotationId)) {
          socket.leave(`org:${user.orgId}:quote:${quotationId}`);
        }
      });
    }

    socket.on('disconnect', (reason: string) => {
      logger.debug({ socketId: socket.id, orgId: user.orgId, reason }, 'Socket disconnected');
    });
  });

  return io;
}

export function getIO(): SocketIOServer {
  if (!io) {
    throw new Error('Socket.IO has not been initialized. Call initSocketIO(httpServer) first.');
  }
  return io;
}

export function emitToOrg(orgId: string, event: string, data: any): void {
  if (!io) return;
  io.to(`org:${orgId}`).emit(event, data);
}

export function emitToQuote(orgId: string, quotationId: string, event: string, data: any): void {
  if (!io) return;
  io.to(`org:${orgId}:quote:${quotationId}`).emit(event, data);
}

File: apps/api-py/src/lib/socket.py
from typing import Any, Optional, Set
import socketio
from src.config.env import settings
from src.lib.logger import logger
from src.shared.jwt_utils import verify_jwt

sio = socketio.AsyncServer(
    async_mode="asgi",
    cors_allowed_origins=settings.cors_origins or "*",
)

@sio.event
async def connect(sid: str, environ: dict, auth: Optional[dict] = None):
    token = None
    if auth and isinstance(auth, dict) and auth.get("token"):
        token = auth["token"]
    elif "HTTP_AUTHORIZATION" in environ and environ["HTTP_AUTHORIZATION"].startswith("Bearer "):
        token = environ["HTTP_AUTHORIZATION"][7:]

    if not token:
        logger.warning(f"Socket connection rejected (sid: {sid}): Missing auth token")
        return False

    try:
        payload = verify_jwt(token)
    except Exception as e:
        logger.warning(f"Socket connection rejected (sid: {sid}): {e}")
        return False

    org_id = payload.get("org_id")
    if not org_id:
        return False

    typ = payload.get("typ")
    role = payload.get("role")
    if typ == "super_admin" or role == "super_admin":
        return False

    if typ == "customer" and role:
        return False

    is_internal = typ == "internal" or (typ is None and role is not None)
    quotation_ids = payload.get("quotation_ids", []) if not is_internal else []

    await sio.save_session(sid, {
        "user_id": payload.get("sub"),
        "org_id": org_id,
        "role": role,
        "is_internal": is_internal,
        "quotation_ids": quotation_ids,
    })

    if is_internal:
        await sio.enter_room(sid, f"org:{org_id}")
        logger.info(f"Internal socket {sid} joined org:{org_id}")
    else:
        for qid in quotation_ids:
            await sio.enter_room(sid, f"org:{org_id}:quote:{qid}")
        logger.info(f"Customer socket {sid} joined {len(quotation_ids)} quote rooms")

    return True

@sio.event
async def disconnect(sid: str):
    logger.debug(f"Socket disconnected {sid}")

@sio.on("quote:join")
async def on_quote_join(sid: str, quotation_id: str):
    session = await sio.get_session(sid)
    if not session:
        return
    org_id = session.get("org_id")
    if not org_id or not isinstance(quotation_id, str):
        return

    quotation_id = quotation_id.strip()
    if not quotation_id:
        return

    if session.get("is_internal"):
        await sio.enter_room(sid, f"org:{org_id}:quote:{quotation_id}")
    else:
        allowed = set(session.get("quotation_ids", []))
        if quotation_id in allowed:
            await sio.enter_room(sid, f"org:{org_id}:quote:{quotation_id}")

@sio.on("quote:leave")
async def on_quote_leave(sid: str, quotation_id: str):
    session = await sio.get_session(sid)
    if not session:
        return
    org_id = session.get("org_id")
    if org_id and isinstance(quotation_id, str):
        await sio.leave_room(sid, f"org:{org_id}:quote:{quotation_id.strip()}")

async def emit_to_org(org_id: str, event: str, data: Any):
    try:
        await sio.emit(event, data, room=f"org:{org_id}")
    except Exception as e:
        logger.warning(f"Failed to emit to org:{org_id} event {event}: {e}")

async def emit_to_quote(org_id: str, quotation_id: str, event: str, data: Any):
    try:
        await sio.emit(event, data, room=f"org:{org_id}:quote:{quotation_id}")
    except Exception as e:
        logger.warning(f"Failed to emit to quote room {quotation_id}: {e}")
